### ISSUE TITLE
Slowdown training LSTM on TensorFlow 2.4+

### ISSUE THREAD

**Issue by zredlined** [tensorflow/tensorflow#44917]
We are noticing a significant slowdown during training with TensorFlow when using the Tensorflow-Privacy LSTM and GRU optimizers on TF 2.4. In tetsing, this only happens when using the TF-privacy optimizers, but it appears related to a change in the recurrent_v2 module of TensorFlow. 

Doing some testing, it looks like the slowdown was introduced in between these two tf-nightly builds.

**Describe the expected behavior**
Training can go from 15 sec/epoch to 2 mins+ per epoch with the latest TF release candidate (tensorflow==2.4.0rc1).
tf-nightly==2.4.0.dev20201019 - 15 sec/epoch

**Describe the current behavior**
tf-nightly==2.4.0.dev20201020 and TensorFlow RC1 - 2 mins+/epoch

**System information**
GCP, running on Tesla V100, 16GB RAM, Ubuntu, 8 vCPU, Python 3.8, cuda11, TensorFlow 2.4.0rc0 and nightly installed via PIP. 

**Standalone code to reproduce the issue**
https://gist.github.com/zredlined/72305ab04670197869e470b232d22ed4

In tensorflow/python/keras/layers/recurrent_v2.py 
I think this TensorFlow commit is the culprit-- changing use_new_code() back to True speeds the code back up. The only reference I can find is in the issue above for what looks like an internal Google issue? Any help would be hugely appreciated, on most datasets we have tested with slowdowns are 10-20x. Thanks!

tensorflow/tensorflow@73b7097.
```
def _use_new_code():
  return False  # NOTE: changed to False in @73b7097. Changing back to True speeds training up.

```

**Other notes/logs**
Originally posted at https://github.com/tensorflow/privacy/issues/141, opening an issue here as it appears to be an issue within TensorFlow.

**Comment by yhliang2018** [comment]
Hi @zredlined, thanks for reporting the issue! Currently, the new code path is disabled due to some other internal issues. In theory, enabling the new code of LSTM/GRU will boost eager performance a lot, but should have little affect to other modes.

Do you run the training with pure eager mode?

**Comment by zredlined** [comment]
Yes, we are running the training in eager mode.

Can you help me understand what the other internal issues might be? We use this for a production library, and need to use TF 2.4.x+ to support differentially private training. Any options to re-enable the codepath (as an optional argument, or whatever you recommend) would be awesome. Thanks!

**Comment by zredlined** [comment]
@amahendrakar Is there any way we can make the codepath for _use_new_code() a user facing option, defaulted to "off" like it is now, but that we can enable to true to speed up eager mode training?

**Comment by misberner** [comment]
@yhliang2018 the training is run via a call to `model.fit` which IIUC should execute the actual training in graph mode. However, the function optimizer is reporting the failure `function_optimizer failed: Invalid argument: Input 0 of node RMSprop/gradients/loop_body/PartitionedCall_1/pfor/PartitionedCall/gradients/zeros_like/pfor/ZerosLike was passed float from sequential/lstm/PartitionedCall:6 incompatible with expected variant.` so some of the graph mode optimizations seem to not get applied. The above optimizer failure message doesn't appear when monkey-patching `_use_new_code()` to return `True`.

**Comment by zredlined** [comment]
+ Andreas Terzis from the TF-Privacy team @aterzis-google

**Comment by yhliang2018** [comment]
@zredlined The new code path with new Case op causes some failures internally, so it's disabled for now. We are working on the fix, and will enabled the new code path eventually. 

In terms of the training mode, as @misberner mentioned, in your repro, `model.compile` with `run_eagerly=None` is used, so the model training with `model.fit` will be wrapped within `tf.function` which is in graph mode. The main change in the new code path is to warp the layer call on CPU into `tf.function`, which in theory only affects pure eager performance on CPU, and should have no effect on graph performance.

I see there is a recent change on `apply_gradients` of `DPKerasAdamOptimizer`, and this might affect model performance as well. @schien1729 any insights into it?

**Comment by yhliang2018** [comment]
> @yhliang2018 the training is run via a call to `model.fit` which IIUC should execute the actual training in graph mode. However, the function optimizer is reporting the failure `function_optimizer failed: Invalid argument: Input 0 of node RMSprop/gradients/loop_body/PartitionedCall_1/pfor/PartitionedCall/gradients/zeros_like/pfor/ZerosLike was passed float from sequential/lstm/PartitionedCall:6 incompatible with expected variant.` so some of the graph mode optimizations seem to not get applied. The above optimizer failure message doesn't appear when monkey-patching `_use_new_code()` to return `True`.

@misberner Where is this error message from? Could you provide a colab link for a mini repro?

**Comment by misberner** [comment]
@yhliang2018 unfortunately I am unable to reproduce this in a Colab notebook even with GPU enabled (or the native library messages are suppressed). I am able to reproduce it on my box by running the code from the gist linked in the original issue w/o any further modifications.
```
$ pip list | grep -E '^(tf-nightly|tensorflow-privacy)'
tensorflow-privacy                 0.5.1
tf-nightly                         2.4.0.dev20201020
$ curl 'https://gist.githubusercontent.com/zredlined/72305ab04670197869e470b232d22ed4/raw/4d9cd113fe38cce4d0b6494595fea143096541d4/tf_privacy_basic.py' -o tf_privacy_basic.py
$ python tf_privacy_basic.py
[...]
WARNING:tensorflow:Converting IndexedSlices(indices=Tensor("gradient_tape/sequential/embedding/embedding_lookup/Reshape_1:0", shape=(1600,), dtype=int32), values=Tensor("gradient_tape/sequential/embedding/embedding_lookup/Reshape:0", shape=(1600, 256), dtype=float32), dense_shape=Tensor("gradient_tape/sequential/embedding/embedding_lookup/VariableShape:0", shape=(2,), dtype=int32)) to a dense representation may make it slow. Alternatively, output the indices and values of the IndexedSlices separately, and handle the vectorized outputs directly.
WARNING:tensorflow:Converting IndexedSlices(indices=Tensor("gradient_tape/sequential/embedding/embedding_lookup/Reshape_1:0", shape=(1600,), dtype=int32), values=Tensor("gradient_tape/sequential/embedding/embedding_lookup/Reshape:0", shape=(1600, 256), dtype=float32), dense_shape=Tensor("gradient_tape/sequential/embedding/embedding_lookup/VariableShape:0", shape=(2,), dtype=int32)) to a dense representation may make it slow. Alternatively, output the indices and values of the IndexedSlices separately, and handle the vectorized outputs directly.
2020-11-19 11:46:07.364349: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:116] None of the MLIR optimization passes are enabled (registered 2)
2020-11-19 11:46:07.428010: I tensorflow/core/platform/profile_utils/cpu_utils.cc:112] CPU Frequency: 2200000000 Hz
2020-11-19 11:46:07.480035: E tensorflow/core/grappler/optimizers/meta_optimizer.cc:592] function_optimizer failed: Invalid argument: Input 0 of node Adam/gradients/loop_body/PartitionedCall/pfor/PartitionedCall/gradients/zeros_like/pfor/ZerosLike was passed float from sequential/lstm/PartitionedCall:6 incompatible with expected variant.
2020-11-19 11:46:07.707187: E tensorflow/core/grappler/optimizers/meta_optimizer.cc:592] function_optimizer failed: Invalid argument: Input 0 of node Adam/gradients/loop_body/PartitionedCall/pfor/PartitionedCall/gradients/zeros_like/pfor/ZerosLike was passed float from sequential/lstm/PartitionedCall:6 incompatible with expected variant.
2020-11-19 11:46:07.795195: W tensorflow/core/common_runtime/process_function_library_runtime.cc:805] Ignoring multi-device function optimization failure: Invalid argument: Input 0 of node Adam/gradients/loop_body/PartitionedCall/pfor/PartitionedCall/gradients/zeros_like/pfor/ZerosLike was passed float from sequential/lstm/PartitionedCall:6 incompatible with expected variant.
```
Note that I do see the first two warnings when running it in Colab as well, but they don't seem to come from native code. Another possibly relevant piece of information is that the errors disappear when running with `CUDA_VISIBLE_DEVICES=-1`.

**Comment by zredlined** [comment]
@amahendrakar can we make the codepath selection for _use_new_code() user selectable and defaulted to "false"?

**Comment by zredlined** [comment]
Is the proposed solution above of making `use_new_code()` user selectable and defaulted to `False` acceptable? Happy to submit a pull request. Thanks!

**Comment by yhliang2018** [comment]
@zredlined Sure, making it as a private flag which users can enable it accordingly seems fine to me. Feel free to submit the PR! Thank you!

**Comment by yhliang2018** [comment]
@misberner Which TF version do you use in your box? Could you try tf-nightly to see if the error still exists? Feel free to open another issue if it still doesn't work.

**Comment by misberner** [comment]
@yhliang2018 this was with `tf-nightly==2.4.0.dev20201020`. I'll try with a more recent nightly and update here. Also note that all 2.4 versions doesn't currently work on Colab with GPU acceleration because of https://github.com/tensorflow/tensorflow/issues/42957

**Comment by yhliang2018** [comment]
@misberner Sounds great. Thanks! Does the error happen only on GPU? Given [#42957](https://github.com/tensorflow/tensorflow/issues/42957), How about trying the latest tf-nightly + CPU?
